### PR TITLE
Improves 'yarn patch'

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

- The current patch mechanism doesn't support modifying a patch multiple times from the same base folder, since the original source isn't available anymore through the Yarn APIs once the patch has been applied (so the locator isn't the same anymore).
- Using `yarn patch` currently requires to manually handle the persistence of the patchfiles, which can often be a little bit frustrating.

**How did you fix it?**
- We now unpack the package twice when calling `yarn patch` - this way `patch-commit` doesn't need to query the Yarn APIs anymore to compute the difference between the two package folders.
- The `yarn patch-commit` command now supports a new `-s,--save` flag that will automatically store the patchfile into a configurable directory (`.yarn/patches` by default), and register it inside the `resolutions` field (so that it can affect both direct and transitive dependencies).

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by clemyan** [comment]
A few comments:

- I think `plugin-init` neess to add `!.yarn/patches` to `.gitignore`
- Can the unpack path be made configurable? I would prefer to make it somewhere under the project (e.g. `.yarn/patches/.workdir/`). That way I can stay within the same editor window and can "go to definition" on deps.

**Comment by arcanis** [comment]
> I think `plugin-init` neess to add `!.yarn/patches` to `.gitignore`

Fixed 👍

> Can the unpack path be made configurable? I would prefer to make it somewhere under the project (e.g. `.yarn/patches/.workdir/`). That way I can stay within the same editor window and can "go to definition" on deps.

That's an interesting idea - I however won't implement it in this PR since it implies a few questions I don't have answers for just yet (when should the folder be garbage collected, should the workdir folder also include the original sources as well, ...).

File: packages/yarnpkg-core/sources/structUtils.ts [hunk 1]
```diff
@@ -615,7 +615,7 @@ export function slugifyIdent(ident: Ident) {
 /**
  * Returns a string from a locator, formatted as a slug (eg. `@types-lodash-npm-1.0.0-abcdef1234`).
  */
-export function slugifyLocator(locator: Locator) {
+export function slugifyLocator(locator: Locator, ext: string = ``) {
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/structUtils.ts:618]: This change seems to be unused

**Comment by clemyan** [comment]
> That's an interesting idea - I however won't implement it in this PR since it implies a few questions I don't have answers for just yet (when should the folder be garbage collected, should the workdir folder also include the original sources as well, ...).

My intention is that it will be used as the temp root instead of `/tmp/`, and if the user sets it then they will be on their own -- Yarn does not manage/cleanup the unpacked folders either way and rely on external mechanisms (e.g. OS).

But now that I typed it out, I realize I can just set `TMP` and get the effect I want.

**Review by merceyz — APPROVED** [review]

**Pull request merged**